plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -479,26 +499,38 @@ traceInside ::
     -> PIRTerm PLC.DefaultUni PLC.DefaultFun
     -> PLC.Type PIR.TyName PLC.DefaultUni ()
     -> PIRTerm PLC.DefaultUni PLC.DefaultFun
-traceInside varName lamName = go
+traceInside varName lamName = go mempty
     where
-        go (LamAbs () n t body) (PLC.TyFun () _dom cod) =
+        go ::
+            Map.Map PLC.TyName (PLCType PLC.DefaultUni)
+            -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+            -> PLCType PLC.DefaultUni
+            -> PIRTerm PLC.DefaultUni PLC.DefaultFun
+        go subst (LamAbs () n t body) (PLC.TyFun () _dom cod) =
         -- when t = \x -> body, => \x -> traceInside body
-            LamAbs () n t (traceInside varName lamName body cod)
-        go LamAbs{} _ =
-            error "traceInside: type mismatched. It should be a function type."
-        go e ty =
+            LamAbs () n t $ go subst body cod
+        go subst (TyAbs () tn1 k body) (PLC.TyForall () tn2 _k ty) =
+        -- when t = /\x -> body, => /\x -> traceInside body
+            -- See Note [Profiling polymorphic functions]
+            let subst' = Map.insert tn2 (PLC.TyVar () tn1) subst
```

Comment:
Are we sure that these entries have unique keys? "If the key is already present in the map, the associated value is replaced with the supplied value."

Reply:
Great question. I think we can: if they don't, that would indicate that we were confusing two different type variables in the uniques that we generate for them, which would cause us problems down the way. Generally I think the plugin makes globally unique names for things. But I'll add a comment.